perun-core/src/test/java/cz/metacentrum/perun/core/impl/AuthzResolverImplIntegrationTest.java

```diff
@@ -78,24 +60,100 @@ public void getRolesOnlyValid() throws Exception {
 		User testUser = perun.getUsersManagerBl().getUserByMember(sess, member1);
 		Group testGroup = setUpGroup(createdVo, member1, "testGroup");
 
+		Vo anotherCreatedVo = perun.getVosManagerBl().createVo(sess, new Vo(2, "another VO", "another VO"));
+
+		AuthzResolver.setRole(sess, testUser, anotherCreatedVo, Role.VOADMIN);
 		AuthzResolver.setRole(sess, testGroup, createdVo, Role.VOADMIN);
 
-		AuthzRoles userRoles = authzResolverImpl.getRoles(testUser);
+		AuthzRoles userRoles = authzResolverImpl.getRoles(testUser, true);
 		assertTrue(userRoles.hasRole(Role.VOADMIN, createdVo));
 
 		perun.getMembersManagerBl().invalidateMember(sess, member1);
 
-		userRoles = authzResolverImpl.getRoles(testUser);
+		userRoles = authzResolverImpl.getRoles(testUser, true);
 		assertFalse(userRoles.hasRole(Role.VOADMIN, createdVo));
+		assertTrue(userRoles.hasRole(Role.VOADMIN, anotherCreatedVo));
 
 		perun.getMembersManagerBl().validateMember(sess,member1);
-		userRoles = authzResolverImpl.getRoles(testUser);
+		userRoles = authzResolverImpl.getRoles(testUser, true);
 		assertTrue(userRoles.hasRole(Role.VOADMIN, createdVo));
+		assertTrue(userRoles.hasRole(Role.VOADMIN, anotherCreatedVo));
 
 		perun.getGroupsManagerBl().expireMemberInGroup(sess, member1, testGroup);
 
-		userRoles = authzResolverImpl.getRoles(testUser);
+		userRoles = authzResolverImpl.getRoles(testUser, true);
 		assertFalse(userRoles.hasRole(Role.VOADMIN, createdVo));
+		assertTrue(userRoles.hasRole(Role.VOADMIN, anotherCreatedVo));
+	}
+
+	@Test
+	public void getUserRolesFromAuthorizedGroupsOnly() throws Exception {
+		System.out.println(CLASS_NAME + "getUserRolesFromAuthorizedGroupsOnly");
+
+		Member testMember = createSomeMember(createdVo);
+		User testUser = perun.getUsersManagerBl().getUserByMember(sess, testMember);
+		Group testGroup = setUpGroup(createdVo, testMember, "testGroup");
+
+		// set role for group
+		AuthzResolver.setRole(sess, testGroup, createdVo, Role.VOADMIN);
+
+		// check that user has VO
+		AuthzRoles userRoles = authzResolverImpl.getRolesObtainedFromAuthorizedGroupMemberships(testUser);
+		assertEquals(1, userRoles.size());
+		assertTrue(userRoles.hasRole(Role.VOADMIN, createdVo));
+	}
+
+	@Test
+	public void getUserRolesWithoutAuthorizedGroupsBased() throws Exception {
+		System.out.println(CLASS_NAME + "getUserRolesWithoutAuthorizedGroupsBased");
+
+		Member testMember = createSomeMember(createdVo);
+		User testUser = perun.getUsersManagerBl().getUserByMember(sess, testMember);
+		Group testGroup = setUpGroup(createdVo, testMember, "testGroup");
+
+		// set role for group
+		AuthzResolver.setRole(sess, testGroup, createdVo, Role.VOADMIN);
+
+		// check that user has VO
+		AuthzRoles userRoles = authzResolverImpl.getRoles(testUser, false);
+		assertFalse(userRoles.hasRole(Role.VOADMIN, createdVo));
+	}
+
+	@Test
+	public void getRoleComplementaryObjectsWithAuthorizedGroupsTest() throws Exception {
```

Comment:
Afterwards an edit to this test to cover multiple authorized groups / different roles would be nice

Reply:
Yes, good suggestion. I have adjusted the test already. Thank you!